src/codegeneration/ParenTrait.js

```diff
@@ -0,0 +1,116 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an 'AS IS' BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  ArgumentList,
+  ArrayLiteralExpression,
+  ExpressionStatement,
+  ParenExpression,
+  VariableDeclaration,
+} from '../syntax/trees/ParseTrees.js';
+import {
+  COMMA_EXPRESSION,
+  FUNCTION_EXPRESSION,
+  OBJECT_LITERAL_EXPRESSION,
+  OBJECT_PATTERN,
+} from '../syntax/trees/ParseTreeType.js';
+
+function wrap(tree) {
+  return new ParenExpression(tree.location, tree);
+}
+
+/**
+ * This function is used as trait to generate a class that that adds parens
+ * around trees as needed.
+ *
+ * Usage:
+ *
+ *  class MyTransformer extends ParenTrait(ParseTreeTransformer) {
+ *    ...
+ *  }
+ *
+ * @param {Function} ParseTreeTransformerClass A class that extends
+ *     ParseTreeTransformer.
+ * @return {Function}
+ */
+export function ParenTrait(ParseTreeTransformerClass) {
+  return class extends ParseTreeTransformerClass {
+    transformVariableDeclaration(tree) {
+      let lvalue = this.transformAny(tree.lvalue);
+      let typeAnnotation = this.transformAny(tree.typeAnnotation);
+      let initializer = this.transformAny(tree.initializer);
+      if (initializer !== null && initializer.type === COMMA_EXPRESSION) {
+        initializer = wrap(initializer);
+      } else if (tree.lvalue === lvalue &&
+          tree.typeAnnotation === typeAnnotation &&
+          tree.initializer === initializer) {
+        return tree;
+      }
+      return new VariableDeclaration(tree.location, lvalue, typeAnnotation,
+                                     initializer);
+    }
+
+    transformExpressionStatement(tree) {
+      let expression = this.transformAny(tree.expression);
+      switch (expression.type) {
+        case OBJECT_LITERAL_EXPRESSION:
+        case OBJECT_PATTERN:
+        case FUNCTION_EXPRESSION:
+          expression = wrap(expression);
+          break;
+      }
+      if (tree.expression === expression) {
+        return tree;
+      }
+      return new ExpressionStatement(tree.location, expression);
+    }
+
+    transformExpressionList_(list) {
+      let expressions = this.transformList(list);
+      let newList = null;
+      for (let i = 0; i < list.length; i++) {
+        let expression = expressions[i];
+        if (expression !== null && expression.type === COMMA_EXPRESSION) {
+          expression = wrap(expression);
+          if (newList === null) {
+            newList = expressions.slice(0, i);
+          }
+          newList.push(expression);
+        } else if (newList !== null) {
+          newList.push(expression);
+        }
```

Comment:
What about the case where the first `expression` does not need to be wrapped?  `newList` is `null` so we seem to drop the `expression`. Can that not happen somehow?